### PR TITLE
feat-kthxbyeSidecar-karma-chart

### PR DESCRIPTION
**Pull request by foxtel-temujincabigao**
- instead of add kthxbye sidecar in alertmanager. add the  kthxbyeSidecar in karma chart. 
- point the alertmanager.uri to alertmanager service and port
-  listen to port 80 as 8080 is already used by karma

### CODE REVIEW

File: charts/karma/templates/deployment.yaml [hunk 1]
```diff
@@ -80,6 +80,20 @@ spec:
               mountPath: /etc/certs/{{ . }}
               readOnly: true
           {{- end }}
+        {{- if .Values.kthxbyeSidecar.enabled }}
+        - name: kthxbye-sidecar
+          image: "{{ .Values.kthxbyeSidecar.image.repository }}:{{ .Values.kthxbyeSidecar.image.tag }}"
+          imagePullPolicy: "{{ .Values.kthxbyeSidecar.image.pullPolicy }}"
+          args:
+            - --listen=:80
```
Comment (desaintmartin) [inline charts/karma/templates/deployment.yaml:88]: Should we listen to another port, non-restricted (> 1024) and not run as root?

File: charts/karma/README.md [hunk 2]
```diff
@@ -100,3 +100,29 @@ existingSecretConfig:
...
 
 **NOTE:** you can either use `existingSecretConfig` or `configMap`, it cannot be both.
+
+
+### kthxbye sidecar
+kthxbye is a tiny daemon that can help with managing short lived acknowledged silences.
+It will continuously extend short lived acknowledgement silences if there are alerts firing against those silences,
+which means that the user doesn't need to worry about setting proper duration for such silences.
+
+```yml
```
Comment (desaintmartin) [inline charts/karma/README.md:110]: Probably not needed to reproduce whole content of values.yaml here.

File: charts/karma/Chart.yaml [hunk 3]
```diff
@@ -6,7 +6,7 @@ home: https://github.com/prymitive/karma
 sources:
   - https://github.com/wiremind/wiremind-helm-charts/tree/main/charts/karma
   - https://github.com/prymitive/karma
-version: 2.0.5
+version: 2.0.6
```
Comment (desaintmartin) [inline charts/karma/Chart.yaml:9]:
```suggestion
version: 2.1.0
```

**Review by desaintmartin — COMMENTED** [review]
Thanks for the PR!

- https://github.com/prymitive/karma/blob/main/docs/CONFIGURATION.md#alert-acknowledgement states that `To use it run an instance of kthxbye with every alertmanager instance or cluster and configure it to use the same comment prefix in comment`, why is it different here? Should we amend upstream documentation?
- Why should we add it as a sidecar instead of a standalone Deployment? Or even different, could we move this to its own chart?
- We should add related Service
- See my other inline notes

File: charts/karma/Chart.yaml (hunk 3 above)
Comment (foxtel-temujincabigao) [reply charts/karma/Chart.yaml]: changed version to 2.1.0

File: charts/karma/README.md (hunk 2 above)
Comment (foxtel-temujincabigao) [reply charts/karma/README.md:110]: shortened the values file in readme

File: charts/karma/templates/deployment.yaml (hunk 1 above)
Comment (foxtel-temujincabigao) [reply charts/karma/templates/deployment.yaml]: yes, I agree. I have replaced the port to 8082. also, have added default nonroot user in the kthxbye sidecar container when securityContext is enabled on the pod level.

**Comment by foxtel-temujincabigao** [comment]
> Thanks for the PR!
> 
>     * https://github.com/prymitive/karma/blob/main/docs/CONFIGURATION.md#alert-acknowledgement states that `To use it run an instance of kthxbye with every alertmanager instance or cluster and configure it to use the same comment prefix in comment`, why is it different here? Should we amend upstream documentation?
> 
>     * Why should we add it as a sidecar instead of a standalone Deployment? Or even different, could we move this to its own chart?
> 
>     * We should add related Service
> 
>     * See my other inline notes

That is correct, the upstream PR recommends to run it as a sidecar of alertmanager. The kthxbye alertmanager.uri defaults to http://localhost:9093 to communicate to alertmanager pod.
However, currently alertmanager helm chart does not support sidecar and I see this as a potential risk on alertmanager in an event of a bug or memoryleak in kthxbye, that could cause alertmanager Evicted or CrashloopBackOff and when that happens the alerting itself would not be alerting. 

If this feature get's merged, indeed we need to update the upstream documentation https://github.com/prymitive/karma/blob/main/docs/CONFIGURATION.md#alert-acknowledgement , informing that kthxbye sidecar can be toggled in karma's chart to extend short lived acknowledgement silences.

I do find it more logical to add it as a sidecar of karma since karma is promoting kthxbye as an extended feature to extend short lived acknowledgement silences. In this way, you can easily toggle kthxbye sidecar in karma's chart should you need that feature. 

At this stage I don't see a separate deployment or chart needed for kthxbye as this is a very tiny daemon.

**Comment by desaintmartin** [comment]
Great! A few last questions:
- in the documentation, it is recommended to put a kthxbye per alertmanager instance or cluster, I suppose your change will not work in the case of several alertmanager instances configured in karma, right? If so, this should be stated in the documentation
- What if we set several replicas of Karma deployment? can't 2 kthxbye instances working on a single alertmanager instance cause issues?

**Comment by foxtel-temujincabigao** [comment]
> 
>     * in the documentation, it is recommended to put a kthxbye per alertmanager instance or cluster, I suppose your change will not work in the case of several alertmanager instances configured in karma, right? If so, this should be stated in the documentation

I have not forseen this, but yes you are correct, it will not work in the case of several alertmanager instances configured in karma since the alertmanager.uri is pointing to a specific alertmanager service. It has to be 1:1 karma:alertmanager relationship if you have multiple alertmanagers in the cluster in this PR. Unless kthxbye supports multiple alertmanager.uri similar to karma configmap in the future. 

TLDR: yes, it should be stated in the documentation that kthxbye sidecar in karma does not support multiple alertmanager configuration. 

> 
>     * What if we set several replicas of Karma deployment? can't 2 kthxbye instances working on a single alertmanager instance cause issues?

I tested scaling karma deployment to 3 replicas, and silence an alert in karma dashboard. I had compared logs to each kthxbye-sidecar instances and observed that once silence has been updated by an instance of kthxbye-sidecar. Other kthxbye-sidecar instance will no longer detect (does not reflect on the logs) the expiring silence, therefore will not be extended multiple occurence. 
It seems that expiring silences becomes a race in between each kthxbye-sidecar instances. However, I do not observe odd behavior and the extention off silences behave as expected. 

```
~ for i in $(kubectl get pods -n kube-system-monitoring --selector=app=karma -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{end}'); 
do echo "pod logs $i";
kubectl logs $i -n kube-system-monitoring -c kthxbye-sidecar | grep 6c506aff-b17c-4cd0-b33f-20bd49362b89;
done
pod logs karma-5c8b85c668-7grkl
pod logs karma-5c8b85c668-c4bw5
level=info msg="Silence updated" id=8ddee49c-9081-475a-81f3-4d0a2e020496 matchers=["alertname=Unavailable Replicas","deployment=~(airflow\\-postgresql|airflow\\-web|dkron|grid\\-scale)","namespace=~(airflow|dkron|selenium\\-hub)","severity=slack","squad=general"] replacedBy=6c506aff-b17c-4cd0-b33f-20bd49362b89
level=info msg="Silence expires in 4m28.341651022s and matches 4 alert(s), extending it by 15m0s" id=6c506aff-b17c-4cd0-b33f-20bd49362b89 matchers=["alertname=Unavailable Replicas","deployment=~(airflow\\-postgresql|airflow\\-web|dkron|grid\\-scale)","namespace=~(airflow|dkron|selenium\\-hub)","severity=slack","squad=general"]
level=info msg="Silence updated" id=6c506aff-b17c-4cd0-b33f-20bd49362b89 matchers=["alertname=Unavailable Replicas","deployment=~(airflow\\-postgresql|airflow\\-web|dkron|grid\\-scale)","namespace=~(airflow|dkron|selenium\\-hub)","severity=slack","squad=general"] replacedBy=280c6daf-f302-4883-a060-16be10cbc37f
pod logs karma-5c8b85c668-f2pq4
```

**Comment by desaintmartin** [comment]
Thanks for the detailed answer! Let's try this, then. One last thing: could you add a test values in the ci directory so that this feature is tested (at least from a syntax & deployment point of view)?

**Comment by foxtel-temujincabigao** [comment]
thanks I have updated the CI test-value.yaml with the neccessary values. enabled should be enough however, other people might have different service name and port.

**Review by desaintmartin — APPROVED** [review]

**Comment by desaintmartin** [comment]
Thanks for the contribution!
